boltkit/controller.py

```diff
@@ -545,3 +600,253 @@ def test():
+            if neo4j_dir.startswith("neo4j"):
+                neo4j_path = path_join(path, dir, cluster_member_dir, neo4j_dir)
+                result = action(neo4j_path)
+                results.append(result)
+                break
+
+    return results
+
+
+def _create_controller(path=None):
+    return WindowsController(path) if platform.system() == "Windows" else UnixController(path)
+
+
+def _extract_http_uri_from_config(config_file_path):
+    with open(config_file_path, "r") as f_in:
+        lines = f_in.readlines()
+
+    http_uri = None
+    for line in lines:
+        if "dbms.connector.http.listen_address" in line:
+            if http_uri is not None:
+                raise RuntimeError("Duplicated http uri configs found in %s" % config_file_path)
+
+            split = line.replace(" ", "").split("dbms.connector.http.listen_address=")
+            raw_uri = split[len(split) - 1]
+            uri = raw_uri if not raw_uri.startswith(":") else localhost(int(raw_uri))
+            http_uri = uri if uri.startswith("http://") else "http://%s" % uri
+
+    parsed_http_uri = urlparse(http_uri)
+    if not parsed_http_uri:
+        raise RuntimeError("Cannot ascertain server address")
+
+    return parsed_http_uri
+
+
+def _extract_cluster_member_info(path):
+    config_file_path = path_join(path, "conf", "neo4j.conf")
+
+    with open(config_file_path, "r") as f_in:
+        lines = f_in.readlines()
+
+    http_uri = None
+    bolt_uri = None
+    for line in lines:
+        if "dbms.connector.http.listen_address" in line:
+            if http_uri is not None:
+                raise RuntimeError("Duplicated http uri configs found in %s" % config_file_path)
+
+            http_uri = _parse_uri("http", line)
+
+        if "dbms.connector.bolt.listen_address" in line:
+            if bolt_uri is not None:
+                raise RuntimeError("Duplicated bolt uri configs found in %s" % config_file_path)
+
+            bolt_uri = _parse_uri("bolt", line)
+
+    return InstanceInfo(http_uri, bolt_uri, path)
+
+
+def _parse_uri(scheme, config_entry):
+    split = config_entry.replace(" ", "").split("=")
+    uri = split[len(split) - 1]
+
+    if uri.startswith(":"):
+        uri = scheme + "://localhost" + uri
+
+    if not uri.startswith(scheme + "://"):
+        uri = scheme + "://" + uri
+
+    parsed_uri = urlparse(uri)
+    if not parsed_uri:
+        raise RuntimeError("Cannot parse uri from config '%s'" % uri)
+
+    return parsed_uri
+
+
+def localhost(port):
+    return "localhost:%d" % port
+
+
+class InstanceInfo:
+    def __init__(self, http_uri, bolt_uri, path):
+        self.http_uri = http_uri
+        self.bolt_uri = bolt_uri
+        self.path = path
+
+    def http_uri_str(self):
+        return self.http_uri.geturl().strip()
+
+    def bolt_uri_str(self):
+        return self.bolt_uri.geturl().strip()
+
+    def __str__(self):
+        return self.http_uri_str() + " " + self.bolt_uri_str() + " " + self.path
```

Comment:
This is used to format `neoctrl-cluster start` output:

```
http://localhost:7474 bolt://localhost:7687 /tmp/cores/core-0/neo4j-enterprise-3.1.0
http://localhost:7475 bolt://localhost:7688 /tmp/cores/core-1/neo4j-enterprise-3.1.0
http://localhost:7476 bolt://localhost:7689 /tmp/cores/core-2/neo4j-enterprise-3.1.0
http://localhost:7477 bolt://localhost:7690 /tmp/read-replicas/read-replica-0/neo4j-enterprise-3.1.0
http://localhost:7478 bolt://localhost:7691 /tmp/read-replicas/read-replica-1/neo4j-enterprise-3.1.0
```

Then client applications have all info about the cluster and can access it. Does this sound sensible?